File: java/syntax/10.Classes.java
// Classes
import java.util.Random;

public class Classes {
    // 1. Basic Class with Constructor and Instance Variables
    static class Person {
        private String name;
        private int age;

        public Person(String name, int age) {
            this.name = name;
            this.age = age;
        }

        public Person(String name) {
            this(name, 36);
        }

        public void introduce() {
            System.out.printf("Hello, my name is %s and I am %d years old.%n", name, age);
        }
    }

    // 2. Class with Class Variables and Static Methods
    static class Employee {
        // Class variable (shared across all instances)
        private static String company = "TechCorp";
        private static int employeeCount = 0;
        private String name;
        private int age;

        public Employee(String name, int age) {
            this.name = name;
            this.age = age;
            employeeCount++;
        }

        public void introduce() {
            System.out.printf("Hello, I'm %s, working at %s%n", name, company);
        }

        public static String getCompanyInfo() {
            return String.format("Company: %s, Total Employees: %d", company, employeeCount);
        }
    }

    // 3. Class with Private Variables and Methods
    static class BankAccount {
        private String accountHolder;
        private double balance;
        private String accountNumber;

        public BankAccount(String accountHolder, double initialBalance) {
            this.accountHolder = accountHolder;
            this.balance = initialBalance;
            this.accountNumber = generateAccountNumber();
        }

        private String generateAccountNumber() {
            Random random = new Random();
            return "ACC" + (10000 + random.nextInt(90000));
        }

        public void deposit(double amount) {
            if (amount > 0) {
                balance += amount;
                System.out.printf("Deposited $%.2f. New balance: $%.2f%n", amount, balance);
            } else {
                System.out.println("Invalid deposit amount");
            }
        }

        public double getBalance() {
            return balance;
        }

        public void introduce() {
            System.out.printf("Account holder: %s, Balance: $%.2f%n", accountHolder, balance);
        }
    }

    // 4. Class with Inheritance
    static class Student extends Person {
        private String grade;
        private String studentId;

        public Student(String name, int age, String grade, String studentId) {
            super(name, age);
            this.grade = grade;
            this.studentId = studentId;
        }

        @Override
        public void introduce() {
            System.out.printf("Hello, I'm %s, a student with ID %s and grade %s%n", 
                name, studentId, grade);
        }

        public void study() {
            System.out.printf("%s is studying hard!%n", name);
        }
    }

    // 5. Class with Property-like behavior
    static class Temperature {
        private double celsius;

        public Temperature(double celsius) {
            setCelsius(celsius);
        }

        public double getCelsius() {
            return celsius;
        }

        public void setCelsius(double value) {
            if (value < -273.15) { // Absolute zero
                throw new IllegalArgumentException("Temperature cannot be below absolute zero");
            }
            this.celsius = value;
        }

        public double getFahrenheit() {
            return (celsius * 9 / 5) + 32;
        }
    }

    public static void main(String[] args) {
        System.out.println("=== Testing Basic Person Class ===");
        Person person1 = new Person("John", 36);
        person1.introduce();

        System.out.println("\n=== Testing Employee Class ===");
        Employee emp1 = new Employee("Alice", 28);
        Employee emp2 = new Employee("Bob", 32);
        emp1.introduce();
        emp2.introduce();
        System.out.println(Employee.getCompanyInfo());

        System.out.println("\n=== Testing BankAccount Class ===");
        BankAccount account = new BankAccount("John Doe", 1000);
        account.deposit(500);
        account.introduce();
        System.out.printf("Current balance: $%.2f%n", account.getBalance());

        System.out.println("\n=== Testing Student Class (Inheritance) ===");
        Student student = new Student("Jane", 20, "A", "STU123");
        student.introduce();
        student.study();

        System.out.println("\n=== Testing Temperature Class (Properties) ===");
        Temperature temp = new Temperature(25);
        System.out.printf("Celsius: %.1f°C%n", temp.getCelsius());
        System.out.printf("Fahrenheit: %.1f°F%n", temp.getFahrenheit());
        temp.setCelsius(30); // Using setter
        System.out.printf("New Celsius: %.1f°C%n", temp.getCelsius());
    }
} 

File: javascript/syntax/10.classes.js
// Classes

// 1. Basic Class with Constructor and Instance Variables
class Person {
    constructor(name, age = 36) {
        this.name = name;
        this.age = age;
    }

    introduce() {
        console.log(`Hello, my name is ${this.name} and I am ${this.age} years old.`);
    }
}

// 2. Class with Class Variables and Static Methods
class Employee {
    // Class variable (shared across all instances)
    static company = "TechCorp";
    static employeeCount = 0;

    constructor(name, age = 30) {
        this.name = name;
        this.age = age;
        Employee.employeeCount++;
    }

    introduce() {
        console.log(`Hello, I'm ${this.name}, working at ${Employee.company}`);
    }

    static getCompanyInfo() {
        return `Company: ${Employee.company}, Total Employees: ${Employee.employeeCount}`;
    }
}

// 3. Class with Private Variables and Methods
class BankAccount {
    #balance;
    #accountNumber;

    constructor(accountHolder, initialBalance = 0) {
        this.accountHolder = accountHolder;
        this.#balance = initialBalance;
        this.#accountNumber = this.#generateAccountNumber();
    }

    #generateAccountNumber() {
        return "ACC" + Math.floor(10000 + Math.random() * 90000);
    }

    deposit(amount) {
        if (amount > 0) {
            this.#balance += amount;
            console.log(`Deposited $${amount}. New balance: $${this.#balance}`);
        } else {
            console.log("Invalid deposit amount");
        }
    }

    getBalance() {
        return this.#balance;
    }

    introduce() {
        console.log(`Account holder: ${this.accountHolder}, Balance: $${this.#balance}`);
    }
}

// 4. Class with Inheritance
class Student extends Person {
    constructor(name, age, grade, studentId) {
        super(name, age); // Call parent constructor
        this.grade = grade;
        this.studentId = studentId;
    }

    introduce() {
        console.log(`Hello, I'm ${this.name}, a student with ID ${this.studentId} and grade ${this.grade}`);
    }

    study() {
        console.log(`${this.name} is studying hard!`);
    }
}

// 5. Class with Property-like behavior
class Temperature {
    #celsius;

    constructor(celsius) {
        this.celsius = celsius; // Using setter
    }

    get celsius() {
        return this.#celsius;
    }

    set celsius(value) {
        if (value < -273.15) { // Absolute zero
            throw new Error("Temperature cannot be below absolute zero");
        }
        this.#celsius = value;
    }

    get fahrenheit() {
        return (this.#celsius * 9 / 5) + 32;
    }
}

// Testing all classes
console.log("=== Testing Basic Person Class ===");
const person1 = new Person("John", 36);
person1.introduce();

console.log("\n=== Testing Employee Class ===");
const emp1 = new Employee("Alice", 28);
const emp2 = new Employee("Bob", 32);
emp1.introduce();
emp2.introduce();
console.log(Employee.getCompanyInfo());

console.log("\n=== Testing BankAccount Class ===");
const account = new BankAccount("John Doe", 1000);
account.deposit(500);
account.introduce();
console.log(`Current balance: $${account.getBalance()}`);

console.log("\n=== Testing Student Class (Inheritance) ===");
const student = new Student("Jane", 20, "A", "STU123");
student.introduce();
student.study();

console.log("\n=== Testing Temperature Class (Properties) ===");
const temp = new Temperature(25);
console.log(`Celsius: ${temp.celsius}°C`);
console.log(`Fahrenheit: ${temp.fahrenheit}°F`);
temp.celsius = 30; // Using setter
console.log(`New Celsius: ${temp.celsius}°C`); 